### ISSUE TITLE
A few LS3MIP missing variables after updating to CMIP6 data request version 01.00.28

### ISSUE THREAD

**Issue by treerink** [EC-Earth/ece2cmor3#394]
The list with new unidentified variables for the LS3MIP specific request (step 3) which popped-up in CMIP6 data request version 01.00.28. This concerns the remaining list of #320:
`````
table  variable     prio  dimensions                         long_name                                    unit        list of MIPs which request this variable

Eday   tsland       1     longitude latitude time            Surface Temperature                          K           LS3MIP                                                                             
Eday   dtes         1     longitude latitude time            Change in surface heat storage               J m-2       LS3MIP                                                                             
Eday   dtesn        1     longitude latitude time            Change in snow/ice cold content              J m-2       LS3MIP                                                                             
Eday   tcs          1     longitude latitude time            Vegetation Canopy Temperature                K           LS3MIP                                                                             
Eday   tgs          1     longitude latitude time            Temperature of bare soil                     K           LS3MIP                                                                             
Eday   albc         1     longitude latitude time            Canopy Albedo                                1           LS3MIP                                                                             
Eday   cnc          1     longitude latitude time            Canopy covered area percentage               %           LS3MIP                                                                             
Eday   es           1     longitude latitude time            Bare soil evaporation                        kg m-2 s-1  LS3MIP                                                                             
Eday   snmsl        1     longitude latitude time            Water flowing out of snowpack                kg m-2 s-1  LS3MIP                                                                             
Eday   dslw         1     longitude latitude time            Change in soil moisture                      kg m-2      LS3MIP                                                                             
Eday   dsn          1     longitude latitude time            Change in snow water equivalent              kg m-2      LS3MIP                                                                             
Eday   dsw          1     longitude latitude time            Change in Surface Water Storage              kg m-2      LS3MIP                                                                             
Eday   dcw          1     longitude latitude time            Change in Interception Storage               kg m-2      LS3MIP                                                                             
Eday   rzwc         1     longitude latitude time            Root zone soil moisture                      kg m-2      LS3MIP                                                                             
Eday   mrsow        1     longitude latitude time            Total Soil Wetness                           1           LS3MIP                                                                             
Eday   mrtws        1     longitude latitude time            Terrestrial Water Storage                    kg m-2      LS3MIP                                                                             
Eday   mrlqso       1     longitude latitude sdepth time     Average layer fraction of liquid moisture    1           LS3MIP                                                                             
Eday   mrfsofr      1     longitude latitude sdepth time     Average layer fraction of frozen moisture    1           LS3MIP                                                                             
Eday   dmlt         1     longitude latitude time stempzero  Depth to soil thaw                           m           LS3MIP                                                                             
Eday   ares         1     longitude latitude time            Aerodynamic resistance                       s m-1       LS3MIP                                                                             
Eday   nudgincsm    1     longitude latitude time            Nudging Increment of Water in Soil Mositure  kg m-2      LS3MIP                                                                             
Eday   nudgincswe   1     longitude latitude time            Nudging Increment of Water in Snow           kg m-2      LS3MIP                                                                             
`````

**Comment by treerink** [comment]
Or if you wish, this list [as a shared google doc xlsx](https://docs.google.com/spreadsheets/d/1ZWeTzGroxLRdUN71zcx9vsAzqmHA5L4_rT069bTGbZs/edit#gid=410667073).

**Comment by treerink** [comment]
Asked Andrea & Franco to have a look.

**Comment by endutra** [comment]
All these variables are not directly output of IFS. Some of them are available in the surface model (HTESSEL), and could be made available in the LS3MIP land_hist simulation. 
Another issue is the variable "es"
Here http://clipc-services.ceda.ac.uk/dreq/u/water_evaporation_flux_from_soil.html it seems that "es" and "evspsblsoi" should be identical ? It that's the case, the current calculation of ece2cmor of evspsblsoi is not correct, and evspsblsoi cannot be provided by ec-earth.

**Comment by etiennesky** [comment]
er are also concerned about the sign of evaporation variables such as evspsblsoi and [evspsbl] Evaporation Including Sublimation and Transpiration - http://clipc-services.ceda.ac.uk/dreq/u/089961a3af4d54d5fb045cf3750e760c.html

In our tests most values are negative (which I assume is correct)?, but in some areas (south of Greenland, around Labrador sea) there are some positive values over the ocean, is that correct?

See this plot  where only positive values are shown in red, and there is a big blob over the Labrador Sea... 

![evspsbl_april_1850](https://user-images.githubusercontent.com/1018556/53250317-7702cd00-36ba-11e9-8ec6-cbd68586cb7f.png)

This is also discussed in the EC-Earth historical monitoric issue.

Example files here, most obvious in the month of April: 


ftp://bscesftp.bsc.es/etourign/CMIP/evspsbl_Amon_EC-Earth3_historical_r2i1p1f1_gr_185001-185012.nc 

ftp://bscesftp.bsc.es/etourign/CMIP/evspsblsoi_Lmon_EC-Earth3_historical_r2i1p1f1_gr_185001-185012.nc

**Comment by endutra** [comment]
The positive evaporation in IFS stands for dew/frost deposition. This happens in reality, but I'm not sure if these high values shown in the plot are realistic ?

**Comment by etiennesky** [comment]
Hi @endutra would you expect dew/frost deposition over the ocean though? I would expect it over the land...

The full discussion is here https://dev.ec-earth.org/issues/610

**Comment by treerink** [comment]
The following two identified variables remain to be implemented, the rest has been added to the ignore list:
`````
Eday	tsland	1  longitude latitude time Surface Temperature  K   Available by IFS: grib code 235 but it needs masking out the  ocean points: thus add in ifspar.json the entry  "masked": "land" Andrea Alessandri       Temperature of the lower boundary of the atmosphere     LS3MIP                                                                                                                  
Eday	mrsow	1  longitude latitude time Total Soil Wetness   1   mrsow=swvl1*0.07+swvl2*0.21+swvl3*0.72+swvl4*1.89  note: NOT divided by maximum allowable soil moisture above wilting point     Andrea Alessandri       Vertically integrated soil moisture divided by maximum allowable soil moisture above wilting point.     LS3MIP                                                                                                                  
`````

**Comment by francocatalano** [comment]
tsland should not be a problem since it is available in IFS (SKT, grib code 235, only ocean masking is needed).

With respect to mrsow, the maximum allowable soil moisture in HTESSEL (soil moisture at saturation) depends on soil texture (constant IFS variable: SLT, code 43). Values are:
-Coarse (SLT=1): 0.403
-Medium (SLT=2): 0.439
-Medium-fine (SLT=3): 0.430
-Fine (SLT=4): 0.520
-Very Fine (SLT=5): 0.614
-Organic (SLT=6): 0.766
-Loamy/Tropical organic (SLT=7): 0.472

Values taken from IFS cycle36r1 manual table 8.8, pag. 123, available here:
https://www.ecmwf.int/file/25465/download?token=NtTZ2HU9

**Comment by treerink** [comment]
In terms of code we need this I guess:
```
# Calculate th_sat (soil moisture at saturation) & th_wilt (soil moisture at wilting point):
if   slt == 1:
 th_sat  = 0.403    # The value for coarse soil
 th_wilt = 0.059    # The value for coarse soil
elif slt == 2:
 th_sat  = 0.439    # The value for medium soil
 th_wilt = 0.151    # The value for medium soil
elif slt == 3:
 th_sat  = 0.430    # The value for medium-fine soil
 th_wilt = 0.133    # The value for medium-fine soil
elif slt == 4:
 th_sat  = 0.520    # The value for fine soil
 th_wilt = 0.279    # The value for fine soil
elif slt == 5:
 th_sat  = 0.614    # The value for very fine soil
 th_wilt = 0.335    # The value for very fine soil
elif slt == 6:
 th_sat  = 0.766    # The value for organic material
 th_wilt = 0.267    # The value for organic material
elif slt == 7:
 th_sat  = 0.472    # The value for loamy/tropical organic material
 th_wilt = 0.171    # The value for loamy/tropical organic material
else:
 print "error"
```

**Comment by treerink** [comment]
I miss definitions of  `swvl1`,` swvl2`,` swvl3`, `swvl4`.

**Comment by francocatalano** [comment]
swvl1 (grib code 039), swvl2 (grib code 040), swvl3 (grib code 041), swvl4 (grib code 042) are the volumetric soil moisture content at levels 1, 2, 3, 4, respectively. They are already included in the LS3MIP ecearth ppt files.

**Comment by treerink** [comment]
Translating this in expressions for the `ifspar.json` I suggest to use auxiliary variables for the `soil type` `(var43)` dependent `th_sat` & `th_wilt`, which should have then expressions like:
```
th_sat(SLT)  = var44 = 44.129 = (var43==1)*0.403+(var43==2)*0.439+(var43==3)*0.430+(var43==4)*0.520+(var43==5)*0.614+(var43==6)*0.766+(var43==7)*0.472
th_wilt(SLT) = var45 = 45.129 = (var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171
```
both to be used in the expression for `mrsow` which also depends on `var39,` `var40,` `var41` & `var42` by:
```
mrsow = (var39 * 0.07 + var40 * 0.21 + var41 * 0.72 + var42 * 1.89 - var45) / (var44 - var45)
```

**Comment by treerink** [comment]
Question: is `mrsow` again a variable which is defined in four soil layers?

**Comment by treerink** [comment]
Seems to be not the case:
```
        "mrsow": {
            "frequency": "day", 
            "modeling_realm": "land", 
            "standard_name": "volume_fraction_of_condensed_water_in_soil_at_field_capacity", 
            "units": "1", 
            "cell_methods": "area: mean where land time: mean", 
            "cell_measures": "area: areacella", 
            "long_name": "Total Soil Wetness", 
            "comment": "Vertically integrated soil moisture divided by maximum allowable soil moisture above wilting point.", 
            "dimensions": "longitude latitude time", 
            "out_name": "mrsow", 
            "type": "real", 
            "positive": "", 
            "valid_min": "", 
            "valid_max": "", 
            "ok_min_mean_abs": "", 
            "ok_max_mean_abs": ""
        }, 
```

**Comment by francocatalano** [comment]
I confirm that mrsow is just total soil moisture, therefore it is not defined over 4 levels.

**Comment by treerink** [comment]
Okay, without intermediate variables this becomes:

```
 mrsow =  (var39 * 0.07 + var40 * 0.21 + var41 * 0.72 + var42 * 1.89 - ((var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171)) / (((var43==1)*0.403+(var43==2)*0.439+(var43==3)*0.430+(var43==4)*0.520+(var43==5)*0.614+(var43==6)*0.766+(var43==7)*0.472) - ((var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171))",
```

**Comment by francocatalano** [comment]
> Translating this in expressions for the `ifspar.json` I suggest to use auxiliary variables for the `soil type` `(var43)` dependent `th_sat` & `th_wilt`, which should have then expressions like:
> 
> ```
> th_sat(SLT)  = var44 = 44.129 = (var43==1)*0.403+(var43==2)*0.439+(var43==3)*0.430+(var43==4)*0.520+(var43==5)*0.614+(var43==6)*0.766+(var43==7)*0.472
> th_wilt(SLT) = var45 = 45.129 = (var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171
> ```
> 
> both to be used in the expression for `mrsow` which also depends on `var39,` `var40,` `var41` & `var42` by:
> 
> ```
> mrsow = (var39 * 0.07 + var40 * 0.21 + var41 * 0.72 + var42 * 1.89 - var45) / (var44 - var45)
> ```

If the expression "(var43==1)*0.403" is:
1 for var43==1
0 elsewhere
then yes, I think it should work

**Comment by treerink** [comment]
Ready for `genecec` and thereafter for running EC-Earth3 and for testing the cmorising of those results.

**Comment by treerink** [comment]
`Genecec` result uploaded in `r7144-control-output-files`  branch.

Closing issue.

**Comment by treerink** [comment]
The invalid values in `mrsow` have a wrong value of -9e+33 instead of 1e+20. For the rest the field ranges between 0 and 5.1 without obvious artefacts.

**Comment by francocatalano** [comment]
the maximum value of mrsow should be around 1. 5.1 seems quite high and it is worth checking. What are the values of th_sat, th_wilt, var39-42 in the point where mrsow is equal to 5.1?

**Comment by treerink** [comment]
`Eday mrsow`:
![mrsow](https://user-images.githubusercontent.com/18733068/69652565-27e72f80-1072-11ea-8ae7-91fab54973ba.png)

The range appears to be -1 to 5.1, for instance a region around the Sahara is below zero. I can not easily investigate intermediate results, I leave that up to you with cdo or so. You can run it with the `r7144-control-output-files` branch.

We will try to correct the invalid value.

**Comment by francocatalano** [comment]
If you share the netcdf files I can have a look and check. Did you run the cmorization test on our ongoing LS3MIP experiment results on cca, as planned?

**Comment by treerink** [comment]
Hi Franco,

No this is just some basic run, only suited for a technically check. That is why I suggest to run yourself one year EC-Earth* with your current setup / branch, but in combination with the control output files from the `r7144-control-output-files` branch. You can do this by going into the `runtime/classic/ctrl/cmip6-output-control-files/` directory in your branch and then use `svn switch`, like:
```
svn switch https://svn.ec-earth.org/ecearth3/branches/development/2019/r7144-control-output-files/runtime/classic/ctrl/cmip6-output-control-files
```
to get the control output files from the `r7144-control-output-files` branch. In this way you test the fields in your own setup, which I think makes more sense.

**Comment by francocatalano** [comment]
We have discussed with Andrea the issue of uncorrect mrsow ranges. The above formula should be corrected as follows:
mrsow = ((var39 * 0.07 + var40 * 0.21 + var41 * 0.72 + var42 * 1.89)/(0.07+0.21+0.72+1.89)- var45) / (var44 - var45)
When testing it please try to do it on our data on cca so that file sharing is easier and I can check the results. Please also DO NOT add any constrain on min/max values for now in order to check how many negative values we get (if any) and where. Thanks.

**Comment by treerink** [comment]
So this is in a one liner, what you propose:
```
 mrsow =  ((var39 * 0.07 + var40 * 0.21 + var41 * 0.72 + var42 * 1.89) / (0.07+0.21+0.72+1.89) - ((var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171)) / (((var43==1)*0.403+(var43==2)*0.439+(var43==3)*0.430+(var43==4)*0.520+(var43==5)*0.614+(var43==6)*0.766+(var43==7)*0.472) - ((var43==1)*0.059+(var43==2)*0.151+(var43==3)*0.133+(var43==4)*0.279+(var43==5)*0.335+(var43==6)*0.267+(var43==7)*0.171))",
```
Please check. Unfortunately, we can't use intermediate variables in the expressions yet.

**Comment by francocatalano** [comment]
Yes, correct.

**Comment by treerink** [comment]
Results look better (going up to max value of 1), but negative values from about -0.5  for desert like regions are still there:
![mrsow](https://user-images.githubusercontent.com/18733068/69977634-72155880-152b-11ea-9c88-e3f6a956b2f7.png)
The invalid value is not fixed yet as well

This is on my test-all test.

**Comment by francocatalano** [comment]
Hi. Looks reasonable but I'd like to better check the areas with negative values. Where can I get the netcdf? Thanks.

**Comment by treerink** [comment]
Seems that in desert like circumstances in this expression the wilting point contribution can be dominant over all the other terms.

I will mail you the nectdf

**Comment by treerink** [comment]
With:
```
ncap2 -s 'where(mrsow < -10000) mrsow=1.0e+20' mrsow_Eday_EC-Earth3_piControl_r1i1p1f1_gr_19900101-19901231.nc ~/mrsow_Eday_EC-Earth3_piControl_r1i1p1f1_gr_19900101-19901231-missing-value-reset.nc
```
the wrong missing values are replaced. This shows better the full )negative) range in ncview, being: -1.1569 - 1.

![mrsow](https://user-images.githubusercontent.com/18733068/70038924-58baed80-15b9-11ea-9fe4-5691f0675d5f.png)

The netcdf has been mailed to you.

**Comment by francocatalano** [comment]
Ok. I have checked mrsow variable in the netcdf. I confirm that values are reasonable. Negative values appear where vertically integrated soil moisture content goes below wilting point, which is physically correct. Unless there are explicit requirements in the quality check of cmorized data to have mrsow>=0, the better option is to leave the variable as it is presently computed (correcting the wrong missing values, of course).

**Comment by treerink** [comment]
Solved in a97110d38b2a53ef62a8e06800f7969b7ecb49ac. Test underway.

**Comment by treerink** [comment]
Tested, looks fine.

Closing.
